src/lib/dataEntities/networks.ts

```diff
@@ -288,41 +288,65 @@ export const l2Networks: L2Networks = {
   },
 }
 
-const getNetwork = async (
-  signerOrProviderOrChainID: SignerOrProvider | number,
-  layer: 1 | 2
+const getNetworkByChainID = <TNetwork extends Network>(
+  chainID: number,
+  networks: NetworkList<TNetwork>
 ) => {
-  const chainID = await (async () => {
-    if (typeof signerOrProviderOrChainID === 'number') {
-      return signerOrProviderOrChainID
-    }
-    const provider = SignerProviderUtils.getProviderOrThrow(
-      signerOrProviderOrChainID
-    )
-
-    const { chainId } = await provider.getNetwork()
-    return chainId
-  })()
-
-  const networks = layer === 1 ? l1Networks : l2Networks
   if (networks[chainID]) {
     return networks[chainID]
-  } else {
-    throw new ArbSdkError(`Unrecognized network ${chainID}.`)
   }
+  throw new ArbSdkError(`Unrecognized network ${chainID}.`)
 }
 
-export const getL1Network = (
-  signerOrProviderOrChainID: SignerOrProvider | number
-): Promise<L1Network> => {
-  return getNetwork(signerOrProviderOrChainID, 1) as Promise<L1Network>
+const getChainIDfromSignerOrProvider = async (
+  signerOrProvider: SignerOrProvider
+) => {
+  const provider = SignerProviderUtils.getProviderOrThrow(signerOrProvider)
+  const { chainId } = await provider.getNetwork()
+  return chainId
 }
-export const getL2Network = (
-  signerOrProviderOrChainID: SignerOrProvider | number
-): Promise<L2Network> => {
-  return getNetwork(signerOrProviderOrChainID, 2) as Promise<L2Network>
+
+type NetworkOrPromiseOfNetwork<
+  TNetwork extends Network,
+  TSignerOrProviderOrNumber extends SignerOrProvider | number
+> = TSignerOrProviderOrNumber extends SignerOrProvider
+  ? Promise<TNetwork>
+  : TNetwork
+
+function getNetwork<
+  TNetwork extends Network,
+  TSignerOrProviderOrNumber extends SignerOrProvider | number
+>(
+  signerOrProviderOrChainID: TSignerOrProviderOrNumber,
+  networks: NetworkList<TNetwork>
+): NetworkOrPromiseOfNetwork<TNetwork, TSignerOrProviderOrNumber>
+function getNetwork<
+  TNetwork extends Network,
+  TSignerOrProviderOrNumber extends SignerOrProvider | number
+>(
+  signerOrProviderOrChainID: TSignerOrProviderOrNumber,
+  networks: NetworkList<TNetwork>
+): Promise<TNetwork> | TNetwork {
+  if (typeof signerOrProviderOrChainID === 'number') {
+    return getNetworkByChainID(signerOrProviderOrChainID, networks)
+  }
+  return getChainIDfromSignerOrProvider(signerOrProviderOrChainID).then(
```

Comment:
Should always `await` before returning, even though there's a slight performance hit to doing this. See: https://github.com/goldbergyoni/nodebestpractices/issues/737

For similar stack trace reasons I prefer not to use `then`, in this case if an error was thrown in getNetworkByChainID you wouldnt get a good stack trace even if the return had been awaited.

Reply:
was trying to get it not return a Promise for number type arg
let me look into this....